pipeline/lib/metric.ml

```diff
@@ -0,0 +1,313 @@
+  let pull = Fmt.str "#%d" pull_number in
+  Fmt.str "[%s](%s)" pull url
+
+let benchmarks_url ~repo_id ~worker ~docker_image =
+  let base_url = Repository.frontend_url () in
+  Fmt.str "%s/%s?worker=%s&image=%s" base_url repo_id worker docker_image
+
+let format_changes_message ~changes ~repository ~worker ~docker_image
+    ~main_commit =
+  let pull_number = Repository.pull_number repository |> Option.get in
+  let commit = Repository.commit_hash repository in
+  let repo_id = Repository.info repository in
+  let main_commit = Option.value ~default:"Unknown hash" main_commit in
+  let metrics_url = benchmarks_url ~repo_id ~worker ~docker_image in
+  let header =
+    Fmt.str
+      "%s (%s) changes the [metrics](%s) as follows in comparison to `main` \
+       (%s) when running on `%s (%s)`:"
+      (pull_url ~repo_id ~pull_number)
+      (commit_url ~repo_id ~commit)
+      metrics_url
+      (commit_url ~repo_id ~commit:main_commit)
+      worker docker_image
+  in
+  Fmt.str "%s\n%s" header (format_changes ~changes)
+
+let post_github_comment ~api ~repository ~pull_number ~text =
+  let open Lwt.Infix in
+  let repo_owner = Repository.owner repository in
+  let repo_name = Repository.name repository in
+  let variables =
+    [
+      ("owner", `String repo_owner);
+      ("name", `String repo_name);
+      ("pullNumber", `Int pull_number);
+    ]
+  in
+  let query =
+    {|
+       query ($owner: String!, $name: String!, $pullNumber: Int!) {
+         repository(owner: $owner, name: $name) {
+           pullRequest(number: $pullNumber) {
+             id
+           }
+         }
+       }
+     |}
+  in
+  let ( / ) a b = Yojson.Safe.Util.member b a in
+  let prId =
+    Github.Api.exec_graphql api query ~variables >|= fun json ->
+    let open Yojson.Safe.Util in
+    let repo = json / "data" / "repository" in
+    let id = repo / "pullRequest" / "id" in
+    id |> to_string
+  in
+  let _ =
+    prId >|= fun id ->
+    let variables = [ ("subjectId", `String id); ("body", `String text) ] in
+    let mutation =
+      {|
+         mutation ($body: String!, $subjectId: ID!) {
+           addComment(input:{body: $body, subjectId: $subjectId}) {
+             commentEdge {
+               node {
+                 body
+               }
+             }
+           }
+         }
+       |}
+    in
+    Github.Api.exec_graphql api mutation ~variables >|= fun json ->
+    let open Yojson.Safe.Util in
+    let body =
+      json / "data" / "addComment" / "commentEdge" / "node" / "body"
+      |> to_string
+    in
+    print_endline @@ body
+  in
+  ()
+
+module NotifyGithub = struct
+  type t = {
+    conninfo : string;
+    repository : Repository.t;
+    github_app : Github.App.t option;
+    worker : string;
+    docker_image : string;
+    pull_number : int;
+  }
+
+  let id = "notify-github"
+  let pp h (key, _) = Fmt.pf h "notify-github %s" key
+
+  module Key = struct
+    type t = string
+
+    let digest t : string = t
```

Comment:
From memories, OCurrent uses the key to determine if an output has already been done: I'm still very afraid of spamming PRs, and I think the code currently sends one message per execution (job_id). Could we try it with one comment max per PR? (so that it acts as a warning, but doesn't keep on re-sending similar messages on further commits)

Reply:
Great point! When I looked at your comment, I had a vague feeling that I did think of this, but couldn't find the code for it immediately with the limited context GitHub shows for the comments. (It has been a while since I worked on this :grin: )

I looked at the code again to find that I have [this "hack"](https://github.com/punchagan/current-bench/blob/bbc404de91116ca2f69d3ce7ae1d6825bf3bec7b/pipeline/lib/metric.ml#L292) to make the `job_id` to be associated with a commit. This means we'd only send a link once per commit, but we could send multiple notifications if commits are pushed onto a PR over time. I've seen other tools like code coverage tools work similarly, and I think this shouldn't feel too spammy to our users?

Reply:
Oh right, I missed that! (and thought it was the ocurrent `job_id`) ... Alright let's try it like that then!

Reply:
I've now changed the "job_id" slightly to include the worker and the docker_image, so that we are able to send messages for different worker+docker_image combinations for the same commit_hash.